### PR TITLE
Release1

### PR DESCRIPTION
**Pull request by ishitachawla**
1. Read the repository files
2. Check if README.md is present
    2.1 Check if Example is present
    2.2 Check if Contribution is present
3. Check if CODEOWNERS file is present in .github
4. Tell whether pull request review branch protection has been enabled. When it is enabled, all commits must be made to a non protected branch via a pull request with the required number of approving reviews. This also includes:
    4.1 Tells whether Dismissal of stale pull request approvals when new commits are pushed has been enabled
    4.2 Tells whether Review required from code owners has been enabled.

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+name: 'File Existence and Reader'
```
Comment (kanika1894) [inline action.yml]: Let's use some action name such as "Action validation checks"
Final name we can take Pulkit's inputs later on

File: action.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+description: 'Check existence and read'
+branding:
```
Comment (kanika1894) [inline action.yml:3]: Where does this get reflected?
Please verify with Pulkit wherever PM inputs are required.

File: action.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  icon: 'file-text'
+  color: 'blue'
+inputs:
+  repo_name:
+    description: "Your GitHub repository"
+    default: "Check-Requirements"
```
Comment (kanika1894) [inline action.yml]: Is it necessary to have a default value?

File: index.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  //check brnch protection
+    async function start(){ 
+    try{
+    var secret_token = core.getInput('GITHUB_TOKEN');
+    const octokit = new Octokit({
+      auth: secret_token,
+    });
+    const repository = core.getInput('repo_name');
+    const ownername = core.getInput('repo_owner');
```
Comment (kanika1894) [inline index.js]: Is there a way to get owner name from a repo_name?
It can help reduce number of inputs we ask from the user

File: index.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
+const core = require('@actions/core');
+const github = require('@actions/github');
+const fs = require('fs');
+const filePath = core.getInput('path')
+const { request } = require('@octokit/request');
+const { Octokit } = require("@octokit/core");
+fs.readdir('./', (err, files) => {
```
Comment (kanika1894) [inline index.js]: There should be a function that gets called to do the file checks.
For every requirement, try having a separate function

**Review by kanika1894 — CHANGES_REQUESTED** [review]
Let's have proper folders to structure code better and then can proceed with the file changes.

File: index.js (hunk 4 above)
Comment (ishitachawla) [reply index.js]: The user doesnt have to give any of these they will get extracted on their own from the repo in which workflow file is present

File: src/main.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
+import * as core from '@actions/core';
+import * as github from '@actions/github';
+import * as fs from 'fs';
+import { request } from '@octokit/request';
+import { Octokit } from "@octokit/core";
+fs.readdir('./', (err, files) => {
+  console.log("entered");
```
Comment (kanika1894) [inline src/main.ts]: Seems these were some test changes, remove these

File: package.json [hunk 7]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "Final",
```
Comment (kanika1894) [inline package.json]: Rename required

File: src/main.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+      const includesReadme = files.includes('README.md');
+      if(includesReadme){
+        console.log("Found readme");
+        fs.readFile('./README.md', function (err, data) {
+        //check example
+        if(data.includes('Example'))
+          console.log("found example");
+        else
+          console.log("Example not found");
```
Comment (kanika1894) [inline src/main.ts]: The error messages should be more intuitive. Something like :
"Please add example workflow in the readMe."

File: src/main.ts [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                
+      //check contribution
+        if(data.includes('Contribution'))
+          console.log("found contribution");
```
Comment (kanika1894) [inline src/main.ts]: here as well
Update all messages that will be printed to user

File: src/main.ts [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    function nodemodulesCheck(){
+      var flag=0;
+    fs.readdir('./src',(err, filess ) => {
+      for(let i = 0; i < filess.length; i++){
+        if(getExtension(filess[i]) === "ts"){
+          flag=1;
+          fs.readdir('./', (err, files) => {
+            const includesnm = files.includes('node_modules');
```
Comment (kanika1894) [inline src/main.ts]: use variable names which is more readable

File: src/main.ts [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+          console.log("Code owners present");
+        else
+          console.log("code owners file absent");
+      })
+    }
+
+    function nodemodulesCheck(){
+      var flag=0;
+    fs.readdir('./src',(err, filess ) => {
+      for(let i = 0; i < filess.length; i++){
```
Comment (kanika1894) [inline src/main.ts]: filess?
spell check

File: src/main.ts [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+            if(includesnm)
+              console.log("node_modules present in master");
+            else
+              console.log("node_modules not present in master");
+          })
+          break;
+        }
+      }
+      if(flag===0)
+        console.log("not dot ts");
```
Comment (kanika1894) [inline src/main.ts]: Why is this required?
This check is required to check the presence of node_modules in the right branch, the user need not have a message if it's not a ts action.
Please be cautious of what the output messages are.

File: src/main.ts [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        }); 
+        console.log(result);
+        return result;
+    }
+        catch(err){
+          console.log(err);
+          return "error";
+    }
+        
+          }
```
Comment (kanika1894) [inline src/main.ts:151]: proper indendation required

**Review by kanika1894 — CHANGES_REQUESTED** [review]
The action flow doesn't seem to show what cases are failing.
In the workflow flow, the error scenarios should show as failed.
Can you please verify it's happening?
Also, attach screenshot of sample workflow file and a sample run.

File: package.json (hunk 7 above)
Comment (ishitachawla) [reply package.json]: Should I rename my repository?

Comment (kanika1894) [reply package.json]: Does this require change in the name of the repo?
Ideally, it should have been something like actions-validation.

File: src/main.ts [hunk 14]
```diff
@@ -0,0 +1,110 @@
+import * as core from '@actions/core';
+import * as github from '@actions/github';
+import * as fs from 'fs';
+import { request } from '@octokit/request';
+import { Octokit } from "@octokit/core";
+fs.readdir('./', (err, files) => {
+  if (err)
```
Comment (kanika1894) [inline src/main.ts]: what error is being checked here?

Comment (ishitachawla) [reply src/main.ts]: if the directory is valid

File: src/main.ts [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+          console.log("node_modules folder is not present in master");
+          })
+      break;
+      }
+    }
+  })
+}
+
+
+async function start(branchname: string){
```
Comment (kanika1894) [inline src/main.ts]: Can you please reorder the functions in the file in which they are being called?
Makes the code more readable

**Comment by kanika1894** [comment]
The attached screenshot looks good to me. Can you please verify with @pulkitaggarwl on do we need to show success messages as well? This is in terms of user experience, so better to confirm with him

File: src/main.ts [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+    const ownername = github.context.repo.owner;
+    const result = await octokit.request('GET /repos/{owner}/{repo}/branches/{branch}/protection/required_pull_request_reviews',{
+      repo: repository,
+      owner: ownername,
+      branch: branchname,
+      headers : { Authorization: 'Bearer ' + secret_token
+           }
+    }); 
+    if(result.data.dismiss_stale_reviews===false)
+      core.setFailed("Please enable Dismiss stale pull request approvals when new commits are pushed for "+branchname)
```
Comment (kanika1894) [inline src/main.ts]: I don't think we need this check for now. As per the requirements, just checking whether main and release branches need approval is sufficient.

Comment (kanika1894) [reply src/main.ts]: Please have one function fulfill one requirement. Let's not merge multiple things in one unless they are related.

File: src/main.ts [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+    if(result.data.require_code_owner_reviews===false)
+      core.setFailed("Please enable Require review from Code Owners for "+branchname)
+    if(result.data.dismiss_stale_reviews===true && result.data.require_code_owner_reviews===true)
+      console.log("Require pull request reviews before merging is enabled for "+branchname);
+    }
+    
+  catch(err){
+    console.log(err);
+    return "error";
```
Comment (kanika1894) [inline src/main.ts]: Should it return "error" _(string)_ or the actual err ?

File: src/main.ts [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+  const octokit = new Octokit({
+  auth: secret_token,
+  });
+  const repository = github.context.repo.repo;
+  const ownername = github.context.repo.owner;
+  const result = await octokit.request('GET /repos/{owner}/{repo}/branches',{
+  owner: ownername,
+  repo: repository,
+  headers : { Authorization: 'Bearer ' + secret_token
+
```
Comment (kanika1894) [inline src/main.ts]: remove blank line

File: src/main.ts [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+  }
+  
+  }
+}
+  catch(err){
+    console.log(err);
+    return "error";
+  }       
+}
+
```
Comment (kanika1894) [inline src/main.ts]: Remove extra blank lines

File: src/main.ts [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+      core.setFailed("Please add CODEOWNERS file");
+  })
+}
+
+function nodemodulesCheck(){
+  fs.readdir('./src',(err, filelist ) => {
+  for(let i = 0; i < filelist.length; i++){
+    if(getExtension(filelist[i]) === "ts"){
+      fs.readdir('./', (err, files) => {
+        const includes_nodemodules = files.includes('node_modules');
```
Comment (kanika1894) [inline src/main.ts]: underscore between node and modules in the variable name or can have M capital
(better readability)

File: src/main.ts [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+    codeownerCheck();
+
+    //Check if nodemodules folder is present in master branch for typescript action
+    nodemodulesCheck();
+
+    //check for branch permissions in main
+    start('main');
+
+    //check for nodemodules folder in releases/*
+    releasesfunc();
```
Comment (kanika1894) [inline src/main.ts]: Follow same naming convention for all checks.
_something_check() seems better

File: src/main.ts [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+fs.readdir('./', (err, files) => {
+  if (err)
+    console.log(err);
+  else {
+
+    //Check for example and Contribution in README
+    readmeChecks(files);
+
+    //Check for CODEOWNERS file in .github folder
+    codeownerCheck();
```
Comment (kanika1894) [inline src/main.ts]: Follow camelCase convention for naming variables/function names
example : this can be codeOwnerCheck

follow the same for other as well

File: src/main.ts [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+    //check example
+    if(data.includes('Example'))
+      console.log("Example workflow is present in README");
+    else
+      core.setFailed("Please add Example workflow in README");
+                
+      //check contribution
+    if(data.includes('Contribution'))
+      console.log("Contribution is present in README");
+
```
Comment (kanika1894) [inline src/main.ts]: Use braces for if-else block
and leave blank lines only where necessary

Ensure proper formatting

File: src/main.ts [hunk 24]
```diff
@@ -0,0 +1,171 @@
...
+      const repository = github.context.repo.repo;
+      const ownername = github.context.repo.owner;
+      //Check for example and Contribution in README
+      readmeChecks(files);
+      //Check for CODEOWNERS file in .github folder
+      codeOwnerCheck();
+      //Check if nodemodules folder is present in master branch for typescript action
+      nodeModulesCheck();
+      //check for branch permissions in main
+      branchPermissionCheck('main', repository, ownername, secret_token, octokit);
```
Comment (kanika1894) [inline src/main.ts]: Is there a possibility that some of the actions repo might have "master" instead of main?
It will be good to investigate a bit on this.

**Comment by kanika1894** [comment]
Can you please attach a screenshot of the workflow file code snippet, in PR description?

File: src/main.ts [hunk 25]
```diff
@@ -0,0 +1,169 @@
...
+      codeOwnerCheck();
+      //Check if nodemodules folder is present in master branch for typescript action
+      nodeModulesCheck();
+      //check for branch permissions in main/master and releases/*
+      branchPermissionCheck(repository, ownername, secret_token, octokit);
+      //check for nodemodules folder in releases/*
+      releasesNodeModulesCheck(repository, ownername, secret_token, octokit);
+    }
+  })
+}
```
Comment (kanika1894) [inline src/main.ts:31]: leave one blank line between all functions

**Comment by BALAGA-GAYATRI** [comment]
@ishitachawla As of now I could see all functions residing in main.ts.  You can split them into different files in src and use those functions as imports in main.ts . This improves readability and makes the main.ts more modular.
cc: @kanika1894
